jaeger-core/src/main/java/com/uber/jaeger/propagation/TextMapCodec.java

```diff
@@ -61,39 +71,102 @@ public void inject(SpanContext spanContext, TextMap carrier) {
+   */
   @Override
   public SpanContext extract(TextMap carrier) {
-    SpanContext context = null;
     Map<String, String> baggage = null;
-    String debugId = null;
+    Map<String, String> debugBaggage = null;
+    @Nullable String debugId = null;
+    @Nullable String traceId = null;
     for (Map.Entry<String, String> entry : carrier) {
       // TODO there should be no lower-case here
       String key = entry.getKey().toLowerCase(Locale.ROOT);
       if (key.equals(contextKey)) {
-        context = SpanContext.contextFromString(decodedValue(entry.getValue()));
+        traceId = decodedValue(entry.getValue());
       } else if (key.equals(Constants.DEBUG_ID_HEADER_KEY)) {
         debugId = decodedValue(entry.getValue());
       } else if (key.startsWith(baggagePrefix)) {
         if (baggage == null) {
           baggage = new HashMap<String, String>();
         }
         baggage.put(keys.unprefixedKey(key, baggagePrefix), decodedValue(entry.getValue()));
+      } else if (key.equals(MANUAL_BAGGAGE_HEADER_KEY)) {
+        String decodedDebugBaggageItems = decodedValue(entry.getValue());
+        debugBaggage = getDebugHeadersAsBaggage(decodedDebugBaggageItems);
       }
     }
-    if (context == null) {
-      if (debugId != null) {
-        return SpanContext.withDebugId(debugId);
-      }
+
+    @NotNull SpanContext context;
+    if (debugId == null && traceId == null) {
       return null;
+    } else if (debugId != null) {
+      // Debug ID is always respected regardless of other data
+      baggage = debugBaggage;
+      context = SpanContext.withDebugId(debugId);
+    } else {
+      // debug ID == null + trace ID != null
+      context = SpanContext.contextFromString(traceId);
     }
+
+    // there is a chance that baggage variable is uninitialized
+    // if there is no `uberctx-` or debug baggage.
     if (baggage == null) {
       return context;
     }
+
     return context.withBaggage(baggage);
   }
 
+  /**
+   * Transform string like "k1=v1,k2=v2" into a map like {k1: v1, k2: v2}.
+   *
+   * All items that do not confirm to the aforementioned rule will not be included. In case where
+   * there is no baggage in the passed string, a null value will be returned.
+   *
+   * @param decodedDebugBaggageItems Contents of debug baggage values in the form "k1=v1,k2=v2"
+   * @return Map of debug headers
+   */
+  @VisibleForTesting protected @Nullable Map<String, String> getDebugHeadersAsBaggage(
+      @NotNull String decodedDebugBaggageItems
+  ) {
+    // strip spaces from both ends to not cause cases like ",,,= " to get picked up
+    decodedDebugBaggageItems = decodedDebugBaggageItems.trim();
+
+    Map<String, String> debugBaggage = null;
+    String[] debugBaggageItems = commaPattern.split(decodedDebugBaggageItems);
+    for (String debugBaggageItem: debugBaggageItems) {
+      String[] debugBaggageItemParts = equalsPattern.split(debugBaggageItem.trim());
```

Comment:
wouldn't a regular string split (non-regex based) be more efficient?

Reply:
actually yes - for shorter strings, a manual split is indeed faster. And I think that's the case we care about. Will make change.